diff --git a/unix/vncpasswd/vncpasswd.cxx b/unix/vncpasswd/vncpasswd.cxx
index 466aa1a2..197d60dc 100644
--- a/unix/vncpasswd/vncpasswd.cxx
+++ b/unix/vncpasswd/vncpasswd.cxx
@@ -147,8 +147,7 @@ static std::vector<uint8_t> readpassword() {
     }
 
     if (first.size() > 8) {
-      fprintf(stderr,"Password should not be greater than 8 characters\nBecause only 8 valid characters are used - try again\n");
-      continue;
+      fprintf(stderr,"Password should not be greater than 8 characters\nBecause only 8 valid characters are used\n");
     }
 
 #ifdef HAVE_PWQUALITY
